### PR TITLE
MOD-2454: remove aiostream dependency: async_merge

### PR DESCRIPTION
**Pull request by kramstrom**
continue splitting https://github.com/modal-labs/modal-client/pull/2320 up into multiple PRs as it grew quite big. This adds `async_merge`. Also ran integration tests that passed

### CODE REVIEW

File: modal/functions.py [hunk 1]
```diff
@@ -206,11 +208,10 @@ async def poll_function(self, timeout: Optional[float] = None):
 
     async def run_generator(self):
         data_stream = _stream_function_call_data(self.client, self.function_call_id, variant="data_out")
```
Comment (freider) [inline modal/functions.py:210]: `_stream_function_call_data` could benefit from a `aclosing` context manager 😬

File: modal/parallel_map.py [hunk 2]
```diff
@@ -249,9 +250,7 @@ async def poll_outputs():
 
         assert len(received_outputs) == 0
 
-    response_gen = stream.merge(drain_input_generator(), pump_inputs(), poll_outputs())
-
-    async with response_gen.stream() as streamer:
+    async with aclosing(async_merge(drain_input_generator(), pump_inputs(), poll_outputs())) as streamer:
```
Comment (freider) [inline modal/parallel_map.py:253]: Same here, the inputs to async_merge would all benefit from `aclosing` to make sure we close those generators in a timely fashion

File: test/async_utils_test.py [hunk 3]
```diff
@@ -382,3 +384,71 @@ async def gen2():
...
+
+@pytest.mark.asyncio
+async def test_async_merge():
+    result = []
+    states = []
+
+    async def gen1():
+        states.append("gen1 enter")
+        try:
+            await asyncio.sleep(0.1)
```
Comment (freider) [inline test/async_utils_test.py:397]: There is a small risk that these sleeps will flake out in CI where things run *sloooow* - not sure there is an easy way to implement the interleaving of generators using Events/Locks instead?

File: modal/_utils/async_utils.py [hunk 4]
```diff
@@ -510,3 +510,37 @@ async def async_zip(*inputs: Union[AsyncIterable[T], Iterable[T]]) -> AsyncGener
             yield tuple(items)
         except StopAsyncIteration:
             break
+
+
+async def async_merge(*inputs: Union[AsyncIterable[T], Iterable[T]]) -> AsyncGenerator[T, None]:
+    queue: asyncio.Queue[Tuple[int, Union[T, Exception]]] = asyncio.Queue()
+
+    async def producer(index: int, iterable: Union[AsyncIterable[T], Iterable[T]]):
```
Comment (freider) [inline modal/_utils/async_utils.py:518]: nit: `index` could be `producer_id` and `active_inputs` could be `active_producers` to make the intention a bit clearer (index could be misunderstood to have to do with the index of an item in the input iterator etc.)

File: test/async_utils_test.py (hunk 3 above)
Comment (kramstrom) [reply test/async_utils_test.py]: changed to events & deterministic output order!

**Comment by kramstrom** [comment]
@prbot autoapprove

**Pull request merged**